server/app/auth/CiviFormProfileMerger.java

```diff
@@ -0,0 +1,77 @@
+package auth;
+
+import java.util.Optional;
+import java.util.function.BiFunction;
+import javax.inject.Provider;
+import models.Account;
+import models.Applicant;
+import org.pac4j.core.profile.UserProfile;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import repository.UserRepository;
+
+/** This class serves to abstract out common user profile merging logic into a shared helper. */
+public class CiviFormProfileMerger {
+
+  private static final Logger logger = LoggerFactory.getLogger(CiviFormProfileMerger.class);
+
+  private final ProfileFactory profileFactory;
+  private final Provider<UserRepository> applicantRepositoryProvider;
+
+  public CiviFormProfileMerger(
+      ProfileFactory profileFactory, Provider<UserRepository> applicantRepositoryProvider) {
+    this.profileFactory = profileFactory;
+    this.applicantRepositoryProvider = applicantRepositoryProvider;
+  }
+
+  /**
+   * This method handles a three way merge between an existing applicant, an existing guest profile,
+   * and an external profile
+   *
+   * @param existingApplicant a potentially existing applicant in the database
+   * @param existingProfile a guest profile in the browser cookie
+   * @param externalProfile an OIDC account in the callback from the OIDC server
+   * @param mergeFunction a function that merges an external profile into a Civiform profile
+   */
+  public <T> Optional<UserProfile> threeWayMerge(
+      Optional<Applicant> existingApplicant,
+      Optional<CiviFormProfile> existingProfile,
+      T externalProfile,
+      BiFunction<CiviFormProfile, T, UserProfile> mergeFunction) {
+
+    if (existingApplicant.isPresent()) {
+      if (existingProfile.isEmpty()) {
+        // Easy merge case - we have an existing applicant, but no guest profile.
+        // This will be the most common.
+        existingProfile = Optional.of(profileFactory.wrap(existingApplicant.get()));
+      } else {
+        // Merge the two applicants and prefer the newer one.
+        existingProfile = twoWayMerge(existingProfile.get(), existingApplicant.get());
+      }
+    }
+
+    // Now merge in the information sent to us by the OIDC server.
+    return Optional.of(
+        mergeFunction.apply(
+            existingProfile.orElseGet(
+                () -> {
```

Comment:
I think this part is skipping creating the profile from the auth data as in


civiformProfileFromOidcProfile
civiformProfileFromSamlProfile

in the two existing flows.

Reply:
Oh good catch. I'll add that.

Reply:
oh existingProfile vs externalProfile is hard to read, I'd suggest renaming externalProfile.  authenticatorProfile, etc

Reply:
Also, Where are you doing the civiformProfileFromOidcProfile and civiformProfileFromSamlProfile functionality?  

You're removing the only callers of them in this PR

Reply:
You can also dev test OIDC with these steps

https://docs.civiform.us/contributor-guide/developer-guide/authentication-providers#testing

and you should see Accounts.authority_id get set correctly with an iss: and sub: data in the string.

Reply:
> Also, Where are you doing the civiformProfileFromOidcProfile and civiformProfileFromSamlProfile functionality?

The previous logic was this for `civiformProfileFromOidcProfile`

1. Creating an empty profile
2. Merging the authenticator profile into the empty profile.

Since step two of that logic is duplicated, I instead replaced those methods with `createEmptyCiviFormProfile` in the new commit of this PR.